### DISCUSSION TITLE
Is there a way to change the cwd in nvim when yazi exits?

### DISCUSSION THREAD

**Discussion by bertradio** [mikavilpas/yazi.nvim#95] [Ideas]
I would like to be able to use yazi to change directories in nvim... similar to yy in bashrc. In other words, if in yazi I change the directory to /bin, when I exit yazi the cwd in nvim would change to /bin.

**Comment by mikavilpas** [comment]
That's a great idea. I think it could be done if I add a way to know what the last directory was when yazi was quit, and if you then add a custom hook.

It sounds good. I actually started adding this feature (last yazi directory) yesterday because I had an idea for another feature.

My idea was to add an optional keybinding for closing yazi and grepping the selected directory using telescope or whatever the user wants to do.

Let's keep this open.

**Comment by mikavilpas** [comment]
By the way, when you change the cwd in nvim, does it cause issues for your other tools such as LSPs or linters, etc? I typically don't change the cwd in my workflow right now, so I don't know.

**Comment by bertradio** [comment]
I do move around a lot in nvim and it's often convenient to change the cwd.

The yazi site has this code to create a function in .bashrc to do exactly that:

function yy() {
local tmp
tmp="$(mktemp -t "yazi-cwd.XXXXXX")"
yazi "$@" --cwd-file="$tmp"
if cwd="$(cat -- "$tmp")" && [ -n "$cwd" ] && [ "$cwd" != "$PWD" ]; then
cd -- "$cwd" || exit
fi
rm -f -- "$tmp"
}

**Comment by mikavilpas** [comment]
I use the same thing in my shell, actually. It's really useful.

How have you found LSP and other tools to work if you change your cwd in neovim?

**Comment by bertradio** [comment]
I've had no problems at all. i'm not a programmer, but I use nvim for all my text editing, especially markdown. Accessing yazi from within nvim has been a big help to me.

**Comment by mikavilpas** [comment]
Ok, I have something that might work for you in the PR autolinked above.

Here's how I set it up in my config:

```lua
{
    -- "mikavilpas/yazi.nvim",
    dir = "~/git/yazi.nvim/",
    dependencies = {
      "nvim-lua/plenary.nvim",
    },
    event = "VeryLazy",
    keys = {
      {
        "<up>",
        function()
          require("yazi").yazi()
        end,
        desc = "Open the file manager",
      },
      {
        "<s-up>",
        function()
          require("yazi").yazi(nil, vim.fn.getcwd())
        end,
        desc = "Open the file manager in the cwd",
      },
    },
    ---@type YaziConfig
    opts = {
      open_for_directories = true,
      enable_mouse_support = true,
      -- log_level = vim.log.levels.DEBUG,
      ---@diagnostic disable-next-line: missing-fields
      hooks = {
        yazi_closed_successfully = function(chosen_file, config, state)
          vim.notify(vim.inspect({ chosen_file, state.last_directory.filename }))
        end,
      },
    },
  }
```

Basically, you can use the `yazi_closed_successfully` hook. I added a new parameter for information that is known when yazi has closed.

Do you know how to try that out? You can add `branch = "cd-events"` to your lazy.nvim spec if you're using that.

**Comment by bertradio** [comment]
Attached is my yazi config. When I enter yazi and change directory then exit it does not change the cwd. Instead I get this notification.
.{
  [2] = "/home/bert/.config/nvim"
}

**Comment by bertradio** [comment]
The directory in the notification is the directory to which I have changed in yazi.

**Comment by mikavilpas** [comment]
Ok, it seems to be working then.
Try this:

```lua
      hooks = {
        yazi_closed_successfully = function(chosen_file, config, state)
          if state.last_directory.filename then
            vim.notify("Changing directory to " .. state.last_directory.filename)
            vim.fn.chdir(state.last_directory.filename)
          end
        end,
      },
```

You can remove the notifying once you have made sure it works.

**Comment by bertradio** [comment]
Works now. Cool!

Only issue is that it changes cwd if I open a file from yazi. Not sure if this is what I want. I do want it to change directory if I exit yazi without picking a file. What behavior do you think is best?

**Comment by bertradio** [comment]
I've played with this awhile and I think the behavior I would like is to change CWD if I exit yazi without opening a file. If I open a file and yazi exits, I would not want to change the cwd. Does this make sense? Is it an easy change?

**Comment by mikavilpas** [comment]
Ah, good point! I think this does what you want:

```lua
      ---@diagnostic disable-next-line: missing-fields
      hooks = {
        yazi_closed_successfully = function(chosen_file, config, state)
          if chosen_file == nil and state.last_directory.filename then
            vim.notify("Changing directory to " .. state.last_directory.filename)
            vim.fn.chdir(state.last_directory.filename)
          end
        end,
      },
```

I am thinking this could be a built in feature in yazi.nvim, but I'm not sure how I could make it happen just yet. If I come up with a way, I'll notify you via this issue, but for now that should work!

**Comment by bertradio** [comment]
Perfect!

How about a setting on opt for this with these options for no cwd change, cwd change always, cwd change except on file open?

**Comment by bertradio** [comment]
Another idea, pass the options when yazi is called. That way one could have different keymaps for different options.

**Comment by mikavilpas** [comment]
Those are good suggestions, thanks. I'd like to keep this open for now. They can probably be implemented, but so many things are open about the near future in terms of possibilities that I don't know what to do yet.

**Comment by bertradio** [comment]
Thank you so much for what you've already done. Works great for me. As for the rest, not critical... maybe some day/

**Comment by elsesiy** [comment]
Hi @mikavilpas, thank you so much for the great plugin. I'm using it every day!

Commenting here since the `yazi_closed_successfully` hook described above was working well until I recently updated from 8bd164d to 5199400 when it stopped and is now behaving very odd.

```
# lazy install config
	{
		"mikavilpas/yazi.nvim",
		dependencies = {
			"nvim-lua/plenary.nvim",
		},
		event = "VeryLazy",
		keys = {
			{
				"-",
				function()
					require("yazi").yazi()
				end,
				desc = "Open the file manager",
			},
			{
				"<leader>cw",
				function()
					require("yazi").yazi(nil, vim.fn.getcwd())
				end,
				desc = "Open the file manager in nvim's working directory",
			},
		},
		opts = {
			open_for_directories = true,
			hooks = {
				--- @diagnostic disable: unused-local
				yazi_closed_successfully = function(chosen_file, config, state)
					-- changes cwd when navigating in yazi without selecting a file
					if chosen_file == nil and state.last_directory.filename then
						vim.fn.chdir(state.last_directory.filename)
					end
				end,
			},
		},
	},
```

An arbitrary example using the following folder structure:
```
|- GitHub
   |- mydir.git (worktree)
       |- master
           |- adir
              |- afile
```

Issue 1:
Opening vim in `mydir.git` and navigating to `afile` does not change the cwd to `adir` as expected any longer but to the parent of `mydir.git` which is `GitHub`. I confirmed this by adding a simple `print("last dir " .. state.last_directory.filename)` which yields `last dir /Users/elsesiy/Developer/GitHub`.

Issue 2:
Once I select `afile`, open yazi via `-` and exit yazi via `q`, the dir is set correctly to `adir`. However, any dir change after that does not change cwd any longer. My guess is that `chose_file` does not reset anymore.

Was there any intended breakage that I missed or is this indeed unexpected behavior?
Thank you for taking a look!

**Reply by mikavilpas** [reply]
I found a bug with the current implementation. I think I was able to fix it here https://github.com/mikavilpas/yazi.nvim/pull/519

Could you try this again when that fix is included? It might not fix everything, but I would like to know if it changes things for you, if possible.

Thanks for the detailed question btw, it's really helpful.

**Reply by elsesiy** [reply]
Thank you for the quick reply @mikavilpas, I tried the patch and it does seem to fix it partially. The behavior is now that once I opened a file in `adir` and reopen yazi and immediately close it, the cwd gets updated. However, it does not work without having opened `afile` prior and it also does not change when just navigating around to a different dir/file without opening it.

Working sequence
1. Open nvim in `mydir.git` (cwd: mydir.git)
2. `-` (open yazi) -> navigate to `adir` -> navigate to `afile` (file enter, in my case via `l`)
3. `-` (open yazi) -> `q` (quit yazi) -> cwd: `adir`

Let me know if there's any further debug information I can provide and thanks again for taking a look!

**Reply by mikavilpas** [reply]
Thanks, this helped me and I found a new bug 🙂 It has been fixed here: https://github.com/mikavilpas/yazi.nvim/pull/523

Could you see if this solves the issue?

**Reply by elsesiy** [reply]
Thank you @mikavilpas, looks like things are back to normal and both issues have been resolved with the latest iteration 🎉
